BiometricLoginKotlin/app/src/main/java/com/example/biometricloginsample/LoginActivity.kt

```diff
@@ -0,0 +1,202 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.example.biometricloginsample
+
+import android.content.Context
+import android.content.Intent
+import android.os.Bundle
+import android.text.Editable
+import android.text.TextWatcher
+import android.util.Log
+import android.view.inputmethod.EditorInfo
+import androidx.appcompat.app.AppCompatActivity
+import androidx.appcompat.widget.AppCompatButton
+import androidx.appcompat.widget.AppCompatEditText
+import androidx.appcompat.widget.AppCompatTextView
+import androidx.biometric.BiometricManager
+import androidx.biometric.BiometricPrompt
+import androidx.lifecycle.Observer
+import androidx.lifecycle.ViewModelProviders
+
+
+/**
+ * 1) after entering "valid" username and password, login button becomes enabled
+ * 2) User clicks biometrics?
+ *   - a) if no template exists, then ask user to register template
+ *   - b) if template exists, ask user to confirm by entering username & password
+ */
+class MainActivity : AppCompatActivity() {
+    val TAG = "MainActivity"
+
+    private lateinit var loginWithPasswordViewModel: LoginWithPasswordViewModel
+
+    private lateinit var biometricPrompt: BiometricPrompt
+    private lateinit var cryptographyManager: CryptographyManager
+    private lateinit var encryptedServerTokenWrapper: CiphertextWrapper
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_main)
+        Log.d(TAG, "onCreate is called")
+
+    }
+
+    override fun onResume() {
+        super.onResume()
+        Log.d(TAG, "onResume called")
+        cryptographyManager = CryptographyManager()
+        val ciphertextWrapper = cryptographyManager.getCiphertextWrapperFromSharedPrefs(
+            applicationContext,
+            SHARED_PREFS_FILENAME,
+            Context.MODE_PRIVATE,
+            CIPHERTEXT_WRAPPER
+        )
+
+        if (ciphertextWrapper != null) {
+            if (SampleAppUser.fakeToken == null) {
+                encryptedServerTokenWrapper = ciphertextWrapper
+                showBiometricPromptForDecryption()
+            } else {
+                // The user has already logged in, so proceed to the rest of the app
+                // this is a todo for you, the developer
+                updateApp(getString(R.string.already_signedin))
+            }
+        } else {
+            loginWithPasswordManager()
+        }
+
+        val useBiometricsView = findViewById<AppCompatTextView>(R.id.use_biometrics)
```

Comment:
This should be done in `onCreate`, but it's better to use view binding anyway. :)